### ISSUE TITLE
[15/16/17/18] ORM: KeyError when updating module after removing model with selection field

### ISSUE THREAD

**Issue by alexis-via** [odoo/odoo#179392]
### Odoo Version

- [X] 15.0
- [X] 16.0
- [X] 17.0
- [x] Other (specify)

### Steps to Reproduce

This bug is 100% reproductible on Odoo master, 17, 16, 15 and 14 (I didn't test on older versions). It happens when you have a module that **declare a new model (regular or transient) with a selection field** and, in a new version of the module, that **model is fully removed** from the code of the module.

Here are the steps to reproduce the bug:
1) install the module test_del_wizard_with_selection enclosed on a fresh database. This module depends on "base" ; it adds a wizard (test.wizard.with.selection.field) with a selection field (choice).
2) In the module  test_del_wizard_with_selection, comment the link to ```security/ir.model.access.csv``` in the manifest and comment the content of ```wizards/__init__.py```
3) restart Odoo with -u test_del_wizard_with_selection
=> Odoo crashes on restart.

Here are the logs for Odoo master:
```
2024-09-05 14:28:55,017 66021 INFO master_debugsel odoo.modules.loading: Model test.wizard.with.selection.field is declared but cannot be loaded! (Perhaps a module was partially removed or renamed)-
2024-09-05 14:28:55,023 66021 INFO master_debugsel odoo.addons.base.models.ir_model: Deleting 146@ir.model.access (test_del_wizard_with_selection.access_test_wizard_with_selection_field)-
2024-09-05 14:28:55,037 66021 INFO master_debugsel odoo.models.unlink: User #1 deleted ir.model.data records with IDs: [6331]-
2024-09-05 14:28:55,038 66021 INFO master_debugsel odoo.models.unlink: User #1 deleted ir.model.access records with IDs: [146]-
2024-09-05 14:28:55,038 66021 INFO master_debugsel odoo.addons.base.models.ir_model: Deleting 850@ir.model.fields.selection (test_del_wizard_with_selection.selection__test_wizard_with_selection_field__choice__last)-
2024-09-05 14:28:55,042 66021 WARNING master_debugsel odoo.modules.loading: Transient module states were reset-
2024-09-05 14:28:55,042 66021 ERROR master_debugsel odoo.modules.registry: Failed to load registry-
2024-09-05 14:28:55,042 66021 ERROR master_debugsel odoo.http: Exception during request handling.-
Traceback (most recent call last):
  File "/home/alexis/new_boite/dev/master/odoo/odoo/modules/registry.py", line 102, in __new__
    return cls.registries[db_name]
  File "<decorator-gen-6>", line 2, in __getitem__
  File "/home/alexis/new_boite/dev/master/odoo/odoo/tools/func.py", line 84, in locked
    return func(inst, *args, **kwargs)
  File "/home/alexis/new_boite/dev/master/odoo/odoo/tools/lru.py", line 34, in __getitem__
    a = self.d[obj]
KeyError: 'master_debugsel'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/alexis/new_boite/dev/master/odoo/odoo/http.py", line 2367, in __call__
    response = request._serve_db()
  File "/home/alexis/new_boite/dev/master/odoo/odoo/http.py", line 1872, in _serve_db
    self.registry, cr_readonly = self._open_registry()
  File "/home/alexis/new_boite/dev/master/odoo/odoo/http.py", line 1488, in _open_registry
    registry = Registry(self.db)
  File "/home/alexis/new_boite/dev/master/odoo/odoo/modules/registry.py", line 104, in __new__
    return cls.new(db_name)
  File "<decorator-gen-12>", line 2, in new
  File "/home/alexis/new_boite/dev/master/odoo/odoo/tools/func.py", line 84, in locked
    return func(inst, *args, **kwargs)
  File "/home/alexis/new_boite/dev/master/odoo/odoo/modules/registry.py", line 129, in new
    odoo.modules.load_modules(registry, force_demo, status, update_module)
  File "/home/alexis/new_boite/dev/master/odoo/odoo/modules/loading.py", line 536, in load_modules
    env['ir.model.data']._process_end(processed_modules)
  File "/home/alexis/new_boite/dev/master/odoo/odoo/addons/base/models/ir_model.py", line 2616, in _process_end
    self._process_end_unlink_record(record)
  File "/home/alexis/new_boite/dev/master/odoo/odoo/addons/base/models/ir_model.py", line 2545, in _process_end_unlink_record
    record.unlink()
  File "/home/alexis/new_boite/dev/master/odoo/odoo/addons/base/models/ir_model.py", line 1665, in unlink
    self._process_ondelete()
  File "/home/alexis/new_boite/dev/master/odoo/odoo/addons/base/models/ir_model.py", line 1704, in _process_ondelete
    Model = self.env[selection.field_id.model]
  File "/home/alexis/new_boite/dev/master/odoo/odoo/api.py", line 592, in __getitem__
    return self.registry[model_name](self, (), ())
  File "/home/alexis/new_boite/dev/master/odoo/odoo/modules/registry.py", line 235, in __getitem__
    return self.models[model_name]
KeyError: 'test.wizard.with.selection.field'
```
[test_del_wizard_with_selection.tar.gz](https://github.com/user-attachments/files/16895915/test_del_wizard_with_selection.tar.gz)

**Comment by alexis-via** [comment]
The module I uploaded in my previous comment is for Odoo 17.0. If you want to reproduce the bug on another Odoo version, just change the version in the manifest of the module.

**Comment by gdgellatly** [comment]
@alexis-via I thought I was the only person with this issue. In every pre-migration I have a stanza which is basically DELETE FROM ir_model_fields_selection WHERE field_id IN <removed fields> and then DELETE FROM ir_model_data WHERE <selection_ids>. I never really traced it, but I suspect it is the ORM removing the field before removing the selections

**Comment by StefanRijnhart** [comment]
An attempt to fix it was closed here for no good reason at all: https://github.com/odoo/odoo/pull/104624

**Comment by alexis-via** [comment]
The bug also happens in odoo 18.0 (tested today).
@rco-odoo That's the bug I told you about this afternoon !

**Comment by rco-odoo** [comment]
Fixed from version 16.0 to current version by https://github.com/odoo/odoo/pull/184764.
